### PR TITLE
feat: create the gift app PE-5161

### PR DESCRIPTION
**Pull request by fedellen**

### CODE REVIEW

File: src/utils/getTopUpQuote.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { TopUpRawResponse } from "@ardrive/turbo-sdk";
+import { paymentServiceUrl } from "../constants";
+
+export async function getTopUpQuote(
```
Comment (fedellen) [inline src/utils/getTopUpQuote.ts]: nit: app only needs checkout url. Or maybe even just -- `forwardToCheckoutSession({usdAmount,recipientEmail}): Promise<void>`

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+          node-version: 20.x
+          cache: "yarn"
+
+      - name: Install dependencies
+        run: yarn --immutable --immutable-cache
+
+      - name: Build
+        run: yarn build
+
+      - name: Deploy gh-pages if on prod
```
Comment (fedellen) [inline .github/workflows/deploy.yml:29]: fwiw I learned we can add dev dependency `gh-pages` and add a deploy command like `yarn gh-pages` (but this is working and fine)

File: src/utils/getTopUpQuote.ts (hunk 1 above)
Comment (fedellen) [reply src/utils/getTopUpQuote.ts]: b65095a

File: src/GiftForm.tsx [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    }
+    if (amount < minUSDAmount) {
+      setUsdAmount(minUSDAmount);
+      return;
+    }
+    setUsdAmount(Number(Number(e.target.value).toFixed(2)));
+  };
+
+  const wincForOneGiB = useWincForOneGiB();
+  const debouncedUsdAmount = useDebounce(usdAmount, 500);
```
Comment (fedellen) [inline src/GiftForm.tsx]: nit: factor out `usdAmountDebounceMs` constant

File: src/GiftForm.tsx [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    const amount = Number(e.target.value);
+    if (amount > maxUSDAmount) {
+      setUsdAmount(maxUSDAmount);
+      return;
+    }
+    if (amount < minUSDAmount) {
+      setUsdAmount(minUSDAmount);
+      return;
+    }
+    setUsdAmount(Number(Number(e.target.value).toFixed(2)));
```
Comment (fedellen) [inline src/GiftForm.tsx]: nit: use amount from above

File: src/GiftForm.tsx [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        </label>
+        <textarea
+          className="form-input"
+          id="gift-message"
+          placeholder={
+            "Enter a message to the recipient of the Turbo Credits here"
+          }
+          maxLength={250}
+          required={false}
+        ></textarea>
```
Comment (fedellen) [inline src/GiftForm.tsx]:
```suggestion
        />
```

File: src/hooks/useDebounce.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { useEffect, useState } from "react";
+
+export function useDebounce<T>(value: T, delay?: number): T {
+  const [debouncedValue, setDebouncedValue] = useState<T>(value);
+
+  useEffect(() => {
+    const timer = setTimeout(() => setDebouncedValue(value), delay || 500);
```
Comment (fedellen) [inline src/hooks/useDebounce.ts]: nit: can use default in giftForm.tsx. factor out `defaultDebounceMs`

File: src/hooks/useCreditsForFiat.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+      .catch((err) => {
+        console.error(err);
+        errorCallback(`Error getting credits for USD amount: ${err.message}`);
+      });
+  }, [debouncedUsdAmount, errorCallback]);
+
+  return [
+    winc ? +winc / wincPerCredit : undefined,
+    usdWhenCreditsWereLastUpdatedRef.current,
+  ];
```
Comment (fedellen) [inline src/hooks/useCreditsForFiat.ts:31]: For the reviewer: We keep track of the usd amount when we get prices so we can update both at the same precise render even if the user were to start changing to a new value

File: src/components/RouterPage.tsx [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+export function RouterPage({ page, errorMessage }: RouterPageProps) {
+  return (
+    <>
+      <LogoHeader errorMessage={errorMessage} />
+      {page}
+    </>
+  );
+}
```
Comment (fedellen) [inline src/components/RouterPage.tsx]: for the reviewer: this puts logo and error display onto each page

Comment (fedellen) [reply src/components/RouterPage.tsx]: TODO: add version footer here 🧠

File: src/GiftForm.tsx (hunk 3 above)
Comment (fedellen) [reply src/GiftForm.tsx]: cbeb926

File: src/GiftForm.tsx (hunk 4 above)
Comment (fedellen) [reply src/GiftForm.tsx]: cbeb926

File: src/hooks/useDebounce.ts (hunk 6 above)
Comment (fedellen) [reply src/hooks/useDebounce.ts]: cbeb926

File: src/GiftForm.tsx [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+    forwardToCheckoutSession(usdAmount, recipientEmail);
+  };
+
+  return (
+    <form className="gift-form">
+      <h1>Gift Credits to a friend.</h1>
+
+      <div className="form-section">
+        <label className="form-label">Suggested USD amount</label>
+        <div className="suggested-amount-buttons">
```
Comment (dtfiedler) [inline src/GiftForm.tsx]: nit: center these

File: src/hooks/useCreditsForFiat.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    TurboFactory.unauthenticated(turboConfig)
+      .getWincForFiat({ amount: USD(debouncedUsdAmount), promoCodes: [] })
+      .then(({ winc }) => {
+        usdWhenCreditsWereLastUpdatedRef.current = debouncedUsdAmount;
+        setWinc(winc);
+      })
+      .catch((err) => {
+        console.error(err);
+        errorCallback(`Error getting credits for USD amount: ${err.message}`);
+      });
```
Comment (dtfiedler) [inline src/hooks/useCreditsForFiat.ts:25]: 🚀

File: src/GiftForm.tsx [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+    isEmailHtmlElementValid;
+
+  const handleSubmit = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
+    e.preventDefault();
+
+    if (!canSubmitForm) {
+      return;
+    }
+
+    forwardToCheckoutSession(usdAmount, recipientEmail);
```
Comment (dtfiedler) [inline src/GiftForm.tsx]: .catch() here and show the error

File: src/utils/forwardToCheckoutSession.ts [hunk 12]
```diff
@@ -0,0 +1,18 @@
+import { paymentServiceUrl } from "../constants";
+
+export async function forwardToCheckoutSession(
```
Comment (dtfiedler) [inline src/utils/forwardToCheckoutSession.ts]: i would maybe make this `createCheckoutSessionURL`, return the URL and the caller decides what to do (in this case, forward)

**Review by dtfiedler — APPROVED** [review]
a few minor comments but looking good!

File: src/GiftForm.tsx (hunk 9 above)
Comment (fedellen) [reply src/GiftForm.tsx]: Do you mean center them into the middle or make the buttons fill the space available? I think the latter would like nicer?

**Review by dtfiedler — DISMISSED** [review]

File: src/GiftForm.tsx (hunk 9 above)
Comment (dtfiedler) [reply src/GiftForm.tsx]: yep, the latter

File: src/components/RouterPage.tsx (hunk 8 above)
Comment (dtfiedler) [reply src/components/RouterPage.tsx]: throw a link to our ToS too

Comment (fedellen) [reply src/components/RouterPage.tsx]: f32c254

File: src/GiftForm.tsx (hunk 9 above)
Comment (fedellen) [reply src/GiftForm.tsx]: flex grow! f32c254

File: src/GiftForm.tsx (hunk 11 above)
Comment (fedellen) [reply src/GiftForm.tsx]: f32c254

File: src/utils/forwardToCheckoutSession.ts (hunk 12 above)
Comment (fedellen) [reply src/utils/forwardToCheckoutSession.ts]: f32c254

File: .github/workflows/deploy.yml [hunk 13]
```diff
@@ -24,6 +24,9 @@ jobs:
       - name: Build
         run: yarn build
 
+      - name: Copy index.html to 404.html
+        run: cp ./dist/index.html ./dist/404.html
```
Comment (fedellen) [inline .github/workflows/deploy.yml:27]: for the reviewer -- when GH pages serves static files, if they don't find anything at the target url they will serve 404.html.

they will only serve static file if there is a repo named that with gh pages. like we are at `github.io/turbo-app` and they will show that but if someone goes to `github.io/turbo-app/gift` they show 404 by default.

By adding this dup 404.html we basically create the app there as well as at index. So GH pages serves our 404.html at subpaths which is actually our app 🎉

Comment (dtfiedler) [reply .github/workflows/deploy.yml:27]: we have a workaround for this in the arns-react app - https://github.com/ar-io/arns-react/blob/develop/404.html

Comment (fedellen) [reply .github/workflows/deploy.yml:27]: I was testing a similar script last night but this seems to solve the problem in a simpler way

File: .github/workflows/deploy.yml [hunk 14]
```diff
@@ -0,0 +1,34 @@
+name: Deploy to GitHub Pages
+
+on:
+  push:
+    branches:
+      - prod
+      - PE-5161_gift_app_init
```
Comment (fedellen) [inline .github/workflows/deploy.yml]: TODO: Remove before merge

File: vite.config.ts [hunk 15]
```diff
@@ -5,6 +5,7 @@ import nodePolyfills from "vite-plugin-node-stdlib-browser";
 
 // https://vitejs.dev/config/
 export default defineConfig({
+  base: "/turbo-app/",
```
Comment (fedellen) [inline vite.config.ts]: TODO: We can and should remove this when moving to a subdomain or we'll be at `gift.ardrive.io/turbo-app`

**Review by matibat — DISMISSED** [review]

File: package.json [hunk 16]
```diff
@@ -3,6 +3,7 @@
   "version": "1.0.0",
   "type": "module",
```
Comment (dtfiedler) [inline package.json:4]: consider using `nodenext` in the tsconfig

File: package.json [hunk 17]
```diff
@@ -14,12 +15,14 @@
     "build": "tsc && vite build",
     "lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
     "format": "prettier --write \"src/**/*.{js,jsx,ts,tsx}\"",
-    "preview": "vite preview"
+    "preview": "vite preview",
+    "test": "echo \"TODO: add tests\" && exit 0"
   },
   "dependencies": {
     "@ardrive/turbo-sdk": "^1.1.0",
```
Comment (dtfiedler) [inline package.json:22]: can you test v1.1.1-alpha-1 and see if it removes the need for vite polyfill plugin below?

Comment (fedellen) [reply package.json:22]: https://github.com/ardriveapp/turbo-app/pull/7 -- It did not seem to fix the problem there

**Review by dtfiedler — APPROVED** [review]

File: package.json (hunk 17 above)
Comment (dtfiedler) [reply package.json:22]: all good - will continue validating on that branch

**Pull request merged**